### ISSUE TITLE
Pattern to use pydantic generated code while adding validation rules and changing fields

### ISSUE THREAD

**Issue by TNO-SlaFleur** [flexiblepower/s2-ws-json-python#1]
We have a couple of options in interacting with the pydantic generated code. For this exercise I have manually created an OpenAPI document that we can use for generation purposes for now which may be found [here](https://github.com/flexiblepower/s2-ws-json-python/blob/main/specification/openapi.yml). I have played around with pydantic and I found three options so far:

1. Writing a new layer that can map from and to pydantic models [here](https://github.com/flexiblepower/s2-ws-json-python/blob/main/src/s2wsjson/common/number_range.py). In action [here](https://github.com/flexiblepower/s2-ws-json-python/blob/main/tests/unit/common/number_range_test.py). Very flexible but lots of boilerplate. Breaks IDE autocomplete for constructor keywords. Would be hard to maintain when S2 changes. While S2 right now may have some changes, we expect it to be stable in the long run.
2. Using inheritance to limit the amount of new boilerplate, but allow us to add our own validation logic on top of pydantic [here](https://github.com/flexiblepower/s2-ws-json-python/blob/main/src/s2wsjson/common/number_range_inheritance.py). In action [here](https://github.com/flexiblepower/s2-ws-json-python/blob/main/tests/unit/common/number_range_inheritance_test.py) Limited the amount of boilerplate from 1. but still very flexible. Provides a layer between pydantic code and the functions we expose. Breaks IDE autocomplete for constructor keywords.
3. Inheriting from the pydantic models and using validators to add logic [here](https://github.com/flexiblepower/s2-ws-json-python/blob/main/src/s2wsjson/common/number_range_validator.py) In action [here](https://github.com/flexiblepower/s2-ws-json-python/blob/main/tests/unit/common/number_range_validator_test.py). While this limits the amount of boilerplate significantly, it also leaks the pydantic function calls to outside of this library (leaky abstraction). In the unit tests we are calling the pydantic functions. We also need to change fields later on (e.g. a string -> UUID, references through a str id -> object reference when in the same message etc.) and this option may limit our flexibility.

Curious to your thoughts!

**Comment by victorgarcia98** [comment]
Thanks for creating this repository. It's really well posed out. Interesting ideas!

Some comments on the different options:

## Option 1

* Very flexible
* A lot of boilerplate code
* Not validating on partial updates (could be done by overloading `__setattr__`.)
* Not validating on initialization, only on dump (could be done in the `__post_init__` method).

## Option 2

* I like there's a common Mixin to expose the methods and attributes that we want
* I think we can make the constructor autocomplete work by removing the constructor. (I've tried it out with VSCode and test run fine).

## Option 3

* It's very concise.
* We can include `validate_assignment = True` to enable validation on partial updates.

## Summary

Option 1 is the most flexible, but also requires a lot of coding which will increase the likelihood of introducing bugs and typos. With that in mind, I would use a combination of options 2 & 3, setting `validate_assignment=True` to enable partial update validation. 

```python
class NumberRangeInherit(GenNumberRange, ValidateValuesMixin):
    class Config:
        validate_assignment = True

    @validator('start_of_range')
    def validate_start_of_range(cls, v):
        if v < 0:
            raise ValueError('start_of_range should be >= 0')
        return v

    def validate_values(self) -> bool:
        if self.start_of_range > self.end_of_range:
            raise S2ValidationError(self, 'start_of_range should not be higher than end_of_range')

        return True
```

Please, let me know what do you think :D

**Comment by victorgarcia98** [comment]
Hej Sebaastian!! Cool work.

Regarding the autocompletion of `Timer`, for me it works in VSCode, showing the right Time when I hover over it.

In the class `ValidateValuesMixin`, I'm not sure what extra benefit the method `validate_across_values` gives us over using pydantic validation, e.g for NumberRange:

```python
@root_validator(pre=True)
def validate_start_end_order(cls, v, values, **kwargs):
    if values.get("start_of_range") > values.get("end_of_range"):
        raise S2ValidationError(self, 'start_of_range should not be higher than end_of_range')
```

For exception handling, I think we should move that to the Mixin and maybe wrap `__setattr__` to translate between pydantic exceptions to our own.

Other than that, I looks good to me and I think we should start implementing the models right away :smile:

**Comment by TNO-SlaFleur** [comment]
Hea Victor!

COuple of things:

- When using `@root_validator` or `@validator`, we should use ValueError as defined by pydantic. Pydantic will convert this to a pydantic.ValidationError which we should later convert to an `S2ValidationError` I believe.
- Cheers for the link to the @root_validator! Somehow I missed this usage for `@root_validator` when reading the docs. I do propose to set `pre=False` so we can work with the parsed values instead of the raw values. Lovely now that the error is raised on the `__init__` and not later on `to_json`.
- Wrapping `__setattr__` does not seem to be the complete solution. We also need to catch the exception in the methods at `ValidationValuesMixing` as well as overwriting `pydantic.main.validate_model` (this last one is necessary due to __init__ from BaseModel).  Overriding `__setattr__` through inheritance does not seem possible. Trying it out I noticed we need to switch the inheritance order of any of our classes so the Mixin comes first. However, then we miss the `BaseModel.__setattr__` as it is now the second import. Literally overriding `__setattr__` with a wrapped function on the `BaseModel` class works.  Then this all could work but I do not like it. As we need to overwrite pydantic internals. However, I do not see a better approach. I have updated `ValidateValuesMixin` with the necessary overwrites to it all working. Do you have any ideas on how to improve the wrapping? A list of items to keep in mind that we need to wrap:
  - __init__
  - assignment
  - parse_raw
  
Proposed changes may be found here: https://github.com/flexiblepower/s2-ws-json-python/pull/2
Please review and merge if no comments.

**Comment by TNO-SlaFleur** [comment]
@victorgarcia98 My holiday starts now. I have prepared 2 pull requests for you to review and merge in the mean time. Feel free to pick up where I left off as well as a bunch of messages still need to override their types e.g. ID -> uuid.UUID type, Duration -> TimeDelta, etc. Should be back on the 14th and I will continue wherever we are by then.

**Comment by victorgarcia98** [comment]
I think we can close this thread, the topic of this issue is addressed in #2 and #7. What do you think @TNO-SlaFleur?

**Comment by TNO-SlaFleur** [comment]
Completely agree! That is my expectation as well.
